# sklearn Multiple linear regression and feature selection using standardization

In [12]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()
from sklearn.linear_model import LinearRegression

In [14]:
data = pd.read_csv(r"F:\1.02.+Multiple+linear+regression.csv")
data

SAT  Rand 1,2,3   GPA
0   1714           1  2.40
1   1664           3  2.52
2   1760           3  2.54
3   1685           3  2.74
4   1693           2  2.83
..   ...         ...   ...
79  1936           3  3.71
80  1810           1  3.71
81  1987           3  3.73
82  1962           1  3.76
83  2050           2  3.81

[84 rows x 3 columns]

In [16]:
data.describe()

SAT  Rand 1,2,3        GPA
count    84.000000   84.000000  84.000000
mean   1845.273810    2.059524   3.330238
std     104.530661    0.855192   0.271617
min    1634.000000    1.000000   2.400000
25%    1772.000000    1.000000   3.190000
50%    1846.000000    2.000000   3.380000
75%    1934.000000    3.000000   3.502500
max    2050.000000    3.000000   3.810000

# Create multiple linear regression
# Describe dependent and independent variables

In [18]:
x = data[['SAT','Rand 1,2,3']]
y = data['GPA']  

# Regression itself

In [20]:
reg = LinearRegression()
reg.fit(x,y)

LinearRegression()

In [22]:
reg.coef_

array([ 0.00165354, -0.00826982])

In [24]:
reg.intercept_

0.29603261264909486

# R-squared

In [26]:
reg.score(x,y)

0.40668119528142843

# Adjusted R-squared

In [28]:
r2 = reg.score(x,y)
n = x.shape[0]
p = x.shape[1]
adjusted_r2 = 1-(1-r2)*(n-1)/(n-p-1)
adjusted_r2

0.39203134825134023

In [30]:
from sklearn.feature_selection import f_regression

In [32]:
f_regression(x,y)

(array([56.04804786,  0.17558437]), array([7.19951844e-11, 6.76291372e-01]))

In [34]:
p_value = f_regression(x,y)[1]
p_value

array([7.19951844e-11, 6.76291372e-01])

In [36]:
p_value.round(3)      # 0 - SAT, 0.676 - Rand 1,2,3

array([0.   , 0.676])

# Creating a summary table

In [38]:
reg_summary = pd.DataFrame(data = ['SAT','Rand 1,2,3'], columns = ['Features'])
reg_summary

Features
0         SAT
1  Rand 1,2,3

In [40]:
reg_summary['Coefficients'] = reg.coef_
reg_summary['p_value'] = p_value.round(3)

In [42]:
reg_summary

Features  Coefficients  p_value
0         SAT      0.001654    0.000
1  Rand 1,2,3     -0.008270    0.676

In [44]:
from sklearn.preprocessing import StandardScaler

In [46]:
scaler = StandardScaler()

In [48]:
scaler.fit(x)

StandardScaler()

In [50]:
x_scaled = scaler.transform(x)

In [52]:
x_scaled

array([[-1.26338288, -1.24637147],
       [-1.74458431,  1.10632974],
       [-0.82067757,  1.10632974],
       [-1.54247971,  1.10632974],
       [-1.46548748, -0.07002087],
       [-1.68684014, -1.24637147],
       [-0.78218146, -0.07002087],
       [-0.78218146, -1.24637147],
       [-0.51270866, -0.07002087],
       [ 0.04548499,  1.10632974],
       [-1.06127829,  1.10632974],
       [-0.67631715, -0.07002087],
       [-1.06127829, -1.24637147],
       [-1.28263094,  1.10632974],
       [-0.6955652 , -0.07002087],
       [ 0.25721362, -0.07002087],
       [-0.86879772,  1.10632974],
       [-1.64834403, -0.07002087],
       [-0.03150724,  1.10632974],
       [-0.57045283,  1.10632974],
       [-0.81105355,  1.10632974],
       [-1.18639066,  1.10632974],
       [-1.75420834,  1.10632974],
       [-1.52323165, -1.24637147],
       [ 1.23886453, -1.24637147],
       [-0.18549169, -1.24637147],
       [-0.5608288 , -1.24637147],
       [-0.23361183,  1.10632974],
       [ 1.68156984,

# Regression with scaled feature

In [55]:
reg = LinearRegression()

In [59]:
reg.fit(x_scaled,y)

LinearRegression()

In [63]:
reg.coef_

array([ 0.17181389, -0.00703007])

In [65]:
reg.intercept_

3.330238095238095

# Creating a summary table

In [69]:
reg_summary = pd.DataFrame([['Bias'],['SAT'],['Rand 1,2,3']], columns = ['Features'])
reg_summary['Weight'] = reg.intercept_, reg.coef_[0], reg.coef_[1]
reg_summary


Features    Weight
0        Bias  3.330238
1         SAT  0.171814
2  Rand 1,2,3 -0.007030

# 'SAT is a statistically significant and meaningful predictor of the dependent variable ( GPA). For every 1 unit increase in SAT, GPA increases by 0.172.
# Rand 1,2,3 is not significant and does not contribute to the predictive power of the model — it can probably be removed. For every 1 unit increase in this random variable, GPA slightly decreases by 0.007.